src/java.desktop/windows/classes/sun/awt/windows/ThemeReader.java

```diff
@@ -81,27 +82,36 @@ public static boolean isXPStyleEnabled() {
     }
 
     // this should be called only with writeLock held
-    private static Long getThemeImpl(String widget) {
-        Long theme = widgetToTheme.get(widget);
+    private static Long getThemeImpl(String widget, int dpi) {
+        Long theme = null;
+
+        Map<String, Long> dpiWidgetToTheme = dpiAwareWidgetToTheme.get(dpi);
+
+        if (dpiWidgetToTheme != null) {
+            theme = dpiWidgetToTheme.get(widget);
+        }
```

Comment:
It looks the the entire method could be implemented as a chained called to `computeIfAbsent`:

```java
return dpiAwareWidgetToTheme.computeIfAbsent(dpi, key -> new HashMap<>())
                            .computeIfAbsent(widget, w -> this::reallyOpenTheme);
```

where `reallyOpenTheme` is a new helper method the body of which is that of the `if (theme == null)` block below.

If implementable, it'll make the code shorter and, hopefully, more readable.